---
title: "Detection of consensus peak regions inside a group of experiences using narrowPeak data"
output:
  BiocStyle::html_document:
    toc: true
bibliography: consensusSeekeR.bib
vignette: >
  %\VignetteIndexEntry{Detection of consensus peak regions inside a group of experiences using narrowPeak files}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignettePackage{consensusSeekeR}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

<br />

Authors: Astrid Louise Deschenes, Fabien Claude Lamaze, Pascal Belleau 
and Arnaud Droit <br />
Compilation date: `r Sys.Date()`


## Licensing

This package and the underlying `consensusSeekeR` code are distributed 
under the Artistic license 2.0. You are free to use and redistribute this 
software.


## Introduction

The genome is commonly represented as a linear sequence, split up in 
multiple chromosomes. Genome data, such as genes, nucleosomes or 
single-nucleotide polymorphisms (SNPs) are linked to the genome by 
occupying either a range of 
positions or a single position on the sequence. Genomic related data 
integration is made possible by treating the data as 
ranges on the genome, which acts as a common scaffold [@Lawrence2013]. 
Bioconductor as developed an infrastructure, including packages such as 
`r Biocpkg("GenomicRanges")`,  `r Biocpkg("IRanges")` and 
`r Biocpkg("GenomicFeatures")`, to facilitate 
the integrative statistical analysis of range-based genomic data.

Ranges have a fundamental role in the analysis of experimental data. As 
an example, the peak calling step, in the analysis of ChIP-seq data, generally
generated narrow peaks outputs which include a peak position located inside 
a genomic range. The analysis of gene expression, using RNA-seq data, is based
on counting the alignments overlapping exons which can be translated as
genomic ranges.

## Loading consensusSeekeR package

As with any R package, the `consensusSeekeR` package should first be loaded 
with the following command:

```{r loadingPackage, warning=FALSE, message=FALSE}
library(consensusSeekeR)
```

## Inputs

ChIP-seq combines chromatin immunoprecipitation (ChIP) with massively parallel 
DNA sequencing to identify the binding sites of DNA-associated proteins. The
obtained sequence reads are first mapped to the reference genome of the 
organism used in the experiments. Bindind sites are them dectected by using
software specialized in transcript factor binding sites identification, such 
MACS [@Zhang2008] and PeakRanger [@Feng2011]. Peak
position is defined as a single base pair.

## Using consensusSeekeR on nucleosome position data

Global gene expression patterns are established and maintained by the 
concerted actions of Transcription Factors (TFs) and the proteins that 
constitute chromatin. The key structural element of chromatin is the 
nucleosome, which consists of an octameric histone core wrapped by 146 bps 
of DNA and connected to its neighbor by approximately 10-80 pbs of linker 
DNA (reference TODO).

### Comparing nucleosome positioning results

## Using consensusSeekeR on ChIP-Seq data

Next-generation DNA sequencing coupled with chromatin immunoprecipitation 
(ChIP-seq) has changed the ability to interrogate whole genome protein-DNA 
interactions by enabeling  genomic locatization of histone 
modifications and transcription-factor binding in living cells. Consortium, 
such as [ENCODE]((https://www.encodeproject.org) have developed, and 
constantly upgrading, a set of standards and guidelines for ChIP experiments 
[@Landt2012].


### ChIP-Seq replicates from one experiment

The Encyclopedia of DNA Elements (ENCODE) Consortium is an international 
collaboration of research groups funded by the National Human Genome Research 
Institute. The [ENCODE website](https://www.encodeproject.org) is portal 
giving access to the data generated by the ENCODE Consortium. The amount of 
data gathered is extensive. Moreover, for one specific experiment, more than 
one ChIP-Seq replicate is often available.

The software used to identify transcript factor binding sites generally 
generates a peak position and a narrow region region for each binding site. 
However, it is quite unlikely that the exact peak position is exactly the same 
across replicates. Even more, there is not yet a consensus on how to analyze 
multiple-replicate ChIP-seq samples [@Yang2014].

The `consensusSeekeR` package can be used to identify consensus regions for
two or more replicates ChIP-Seq samples. The consensus regions are being found 
by using the peak positions.

The transcription factor binding for the CTCF transcription factor have been
analyzed and 2 replicates are available in BAM files format on 
[ENCODE website](https://www.encodeproject.org) (DCCs: 
ENCFF000MYJ and ENCFF000MYN). (TODO FABIEN : ajouter une courte description 
sur la methode pour obtenir narrowPeak). To simplify this demo, only 

First, the `consensusSeekeR` package must be loaded.

```{r libraryLoad, warning=FALSE, message=FALSE}
library(consensusSeekeR)
```

### ChIP-Seq data from multiple experiments

The `consensusSeekeR` package can also be used to identify consensus regions 
for two or more ChIP-Seq samples from multiple experiments. While the 
consensus regions are being found by using the peak positions, the TODO.

The transcription factor binding for the CTCF transcription factor have been
analyzed and 2 replicates are available in BAM files format on 
[ENCODE website](https://www.encodeproject.org) (DCCs: 
ENCFF000MYJ and ENCFF000MYN). (TODO FABIEN : ajouter une courte description 
sur la methode pour obtenir narrowPeak).

## Effect of the extendingSize parameter

The value of the `extendingSize` parameter can affect the final number of 
consensus regions. While small `extendingSize` value can miss some regions, 
large `extendingSize` value can gather consensus regions.


## References

