<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{consensusSeekeR: Detection of consensus peak regions inside a group of experiences using narrowPeak files }
-->

<center> <h1>Detection of consensus regions inside a group of experiences 
using narrowPeak data</h1> </center>

<center>Astrid Louise Deschenes, Fabien Claude Lamaze, Pascal Belleau 
and Arnaud Droit.</center>
</br>
<center>This package and the underlying consensusSeekeR code are distributed 
under the Artistic license 2.0.</center> 
<center>You are free to use and redistribute this software.</center>

## Introduction


## Loading the consensusSeekeR package

```{r loadingPackage} 
suppressMessages(library(consensusSeekeR))
```

## Inputs

ChIP-seq combines chromatin immunoprecipitation (ChIP) with massively parallel 
DNA sequencing to identify the binding sites of DNA-associated proteins. 
Sequences are aligned on a reference and bindind sites are dectected by using
software specialized in transcript factor binding sites identification, such 
MACS (Zhang et al., 2008) and PeakRanger (Feng, Grossman et Stein, 2011).

## Using consensusSeekeR on ChIP-Seq data

### ChIP-Seq replicates from one experiment

The Encyclopedia of DNA Elements (ENCODE) Consortium is an international 
collaboration of research groups funded by the National Human Genome Research 
Institute. The [ENCODE website](https://www.encodeproject.org) is portal 
giving access to the data generated by the ENCODE Consortium. The amount of 
data gathered is extensive. Moreover, for one specific experiment, more than 
one ChIP-Seq replicate is often available.

The software used to identify transcript factor binding sites often generates 
a peak position and a narrow region region for each binding site. However, it 
is quite unlikely that the exact peak position is the same across replicates. 

The `consensusSeekeR` package can be use to identify consensus regions for
two or more replicates ChIP-Seq data. The consensus regions are being found by 
using the peak positions.

The transcription factor binding for the FOXA2 transcription factor have been
tested and 2 replicates are available in narrowPeakFile format (DCCs: 
ENCFF001TSH and ENCFF001TSI).



### ChIP-Seq data from multiple experiments

## References

Feng X., Grossman R., Stein L. PeakRanger:A cloud-enabled peak caller for 
ChIP-seq data. BMC Bioinformatics 2011, 12(1):139.

Yang Y, Fear J, Hu J, Haecker I, Zhou L, et al. (2014) Leveraging biological 
replicates to improve analysis in ChIP-seq experiments. Comput Struct 
Biotechnol J 9: e201401002.

Zhang Y., Liu T., Meyer, C.A., Eeckhoute J. Johnson D.S., Bernstein, B.E., 
Nusbaum C., Myer R.M., Brown M., Li W. Model-based Analysis of 
ChIP-Seq (MACS). Genome Biol (2008) vol. 9 (9) pp. R137.
