<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{consensusSeekeR: Detection of consensus peak regions inside a group of experiences using narrowPeak files }
-->

<center> <h1>Detection of consensus regions inside a group of experiences 
using narrowPeak data</h1> </center>

<center>Astrid Louise Deschenes, Fabien Claude Lamaze, Pascal Belleau 
and Arnaud Droit.</center>
</br>
<center>This package and the underlying consensusSeekeR code are distributed 
under the Artistic license 2.0.</center> 
<center>You are free to use and redistribute this software.</center>

## Introduction

Next-generation DNA sequencing coupled with chromatin immunoprecipitation 
(ChIP-seq) has changed our ability to interrogate whole genome protein-DNA 
interactions by enabeling  genomic locatization of histone 
modifications and transcription-factor binding in living cells. Consortium, 
such as [ENCODE]((https://www.encodeproject.org) have developed, and 
constantly upgrading, a set of standards and guidelines for ChIP experiments. 
However, it 



## Loading the consensusSeekeR package

```{r loadingPackage}
suppressMessages(library(consensusSeekeR))
```

## Inputs

ChIP-seq combines chromatin immunoprecipitation (ChIP) with massively parallel 
DNA sequencing to identify the binding sites of DNA-associated proteins. The
obtained sequence reads are first mapped to the reference genome of the 
organism used in the experiments. Bindind sites are them dectected by using
software specialized in transcript factor binding sites identification, such 
MACS (Zhang et al., 2008) and PeakRanger (Feng, Grossman et Stein, 2011). Peak
position is defined as a single base pair.

## Using consensusSeekeR on ChIP-Seq data

### ChIP-Seq replicates 

The Encyclopedia of DNA Elements (ENCODE) Consortium is an international 
collaboration of research groups funded by the National Human Genome Research 
Institute. The [ENCODE website](https://www.encodeproject.org) hold data that 
is freely available for download and analysis.

First, the `consensusSeekeR` package must be loaded.

```{r libraryLoad}
suppressMessages(library(consensusSeekeR))
```

### ChIP-Seq multiple experiments

## Effect of the extendingSize parameter

The value of the `extendingSize` parameter can affect the final number of 
consensus regions. While small `extendingSize` value can miss some regions, 
large `extendingSize` value can gather consensus regions.



## Using consensusSeekeR on ChIP-Seq samples

### Replicate ChIP-Seq samples from one experiment

The Encyclopedia of DNA Elements (ENCODE) Consortium is an international 
collaboration of research groups funded by the National Human Genome Research 
Institute. The [ENCODE website](https://www.encodeproject.org) is portal 
giving access to the data generated by the ENCODE Consortium. The amount of 
data gathered is extensive. Moreover, for one specific experiment, more than 
one ChIP-Seq replicate is often available.

The software used to identify transcript factor binding sites generally 
generates a peak position and a narrow region region for each binding site. 
However, it is quite unlikely that the exact peak position is exactly the same 
across replicates. Finally, there is not yet a consensus on how to analyze 
multiple-replicate ChIP-seq samples (Yang et al., 2008).

The `consensusSeekeR` package can be used to identify consensus regions for
two or more replicates ChIP-Seq samples. The consensus regions are being found 
by using the peak positions.

The transcription factor binding for the CTCF transcription factor have been
analyzed and 2 replicates are available in BAM files format on 
[ENCODE website](https://www.encodeproject.org) (DCCs: 
ENCFF000MYJ and ENCFF000MYN). (TODO FABIEN : ajouter une courte description 
sur la methode pour obtenir narrowPeak). To simplify this demo, only 


### ChIP-Seq samples from multiple experiments

The `consensusSeekeR` package can also be used to identify consensus regions 
for two or more ChIP-Seq samples from multiple experiments. While the 
consensus regions are being found by using the peak positions, the .

The transcription factor binding for the CTCF transcription factor have been
analyzed and 2 replicates are available in BAM files format on 
[ENCODE website](https://www.encodeproject.org) (DCCs: 
ENCFF000MYJ and ENCFF000MYN). (TODO FABIEN : ajouter une courte description 
sur la methode pour obtenir narrowPeak).

## References

Feng X., Grossman R., Stein L. PeakRanger:A cloud-enabled peak caller for 
ChIP-seq data. BMC Bioinformatics 2011, 12(1):139.

Yang Y, Fear J, Hu J, Haecker I, Zhou L, et al. (2014) Leveraging biological 
replicates to improve analysis in ChIP-seq experiments. Comput Struct 
Biotechnol J 9: e201401002.

Zhang Y., Liu T., Meyer, C.A., Eeckhoute J. Johnson D.S., Bernstein, B.E., 
Nusbaum C., Myer R.M., Brown M., Li W. Model-based Analysis of 
ChIP-Seq (MACS). Genome Biol (2008) vol. 9 (9) pp. R137.
